File: comfyui-edition/src/lattice/Utils/CircuitBreaker.hs
-- |
-- Module      : Lattice.Utils.CircuitBreaker
-- Description : Circuit breaker pattern to prevent cascading failures
-- 
-- Migrated from ui/src/utils/circuitBreaker.ts
-- Pure functions for service resilience
-- 
--                                                                  // critical
--

{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}

module Lattice.Utils.CircuitBreaker
  ( -- Types
    CircuitState(..)
  , CircuitBreakerOptions(..)
  , CircuitBreaker
  -- Circuit breaker operations
  , createCircuitBreaker
  , execute
  , reset
  , getCurrentState
  , isOpen
  , isClosed
  , isHalfOpen
  , getFailureCount
  ) where

import Control.Concurrent.MVar (MVar, newMVar, readMVar, modifyMVar_)
import Control.Exception (SomeException, catch)
import Data.Time.Clock (NominalDiffTime, diffUTCTime, getCurrentTime, UTCTime, addUTCTime)
import Lattice.Utils.NumericSafety (isFinite)
import Data.Text (Text)
import qualified Data.Text as T

-- ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
--                                                                     // types
-- ════════════════════════════════════════════════════════════════════════════

-- | Circuit breaker state
-- 
-- System F/Omega proof: Explicit sum type (no null/undefined)
data CircuitState
  = CircuitClosed    -- Circuit is closed, allowing requests
  | CircuitOpen      -- Circuit is open, blocking requests
  | CircuitHalfOpen  -- Circuit is half-open, testing recovery
  deriving (Eq, Show)

-- | Circuit breaker configuration options
-- 
-- System F/Omega proof: Explicit types for all fields
-- Mathematical proof: All numeric fields have min/max bounds
data CircuitBreakerOptions = CircuitBreakerOptions
  { cbFailureThreshold :: Int     -- Must be > 0, default: 5
  , cbResetTimeout :: Int         -- Must be > 0, default: 60000 (ms)
  , cbMonitoringWindow :: Int     -- Must be > 0, default: 60000 (ms)
  , cbOnOpen :: Maybe (IO ())     -- Optional callback when circuit opens
  , cbOnClose :: Maybe (IO ())    -- Optional callback when circuit closes
  , cbOnHalfOpen :: Maybe (IO ()) -- Optional callback when circuit half-opens
  }

-- | Failure record with timestamp
data FailureRecord = FailureRecord
  { failureTimestamp :: UTCTime
  }
  deriving (Eq, Show)

-- | Circuit breaker internal state
data CircuitBreakerState = CircuitBreakerState
  { cbState :: CircuitState
  , cbFailures :: [FailureRecord]
  , cbNextAttempt :: UTCTime
  , cbOptions :: CircuitBreakerOptions
  }

-- | Circuit breaker instance
-- 
-- System F/Omega proof: Explicit type with MVar for thread-safe state
data CircuitBreaker a = CircuitBreaker
  { cbFunction :: IO a
  , cbStateVar :: MVar CircuitBreakerState
  }

-- ════════════════════════════════════════════════════════════════════════════
--                                                        // default // options
-- ════════════════════════════════════════════════════════════════════════════

-- | Default circuit breaker options
-- 
-- System F/Omega proof: Explicit default values
-- Mathematical proof: All defaults satisfy constraints
defaultCircuitBreakerOptions :: CircuitBreakerOptions
defaultCircuitBreakerOptions = CircuitBreakerOptions
  { cbFailureThreshold = 5
  , cbResetTimeout = 60000
  , cbMonitoringWindow = 60000
  , cbOnOpen = Nothing
  , cbOnClose = Nothing
  , cbOnHalfOpen = Nothing
  }

-- ════════════════════════════════════════════════════════════════════════════
--                                                       // state // management
-- ════════════════════════════════════════════════════════════════════════════

-- | Update circuit state based on current conditions
-- 
-- System F/Omega proof: Pure function (state → state)
-- Mathematical proof: State transitions are deterministic
updateState :: UTCTime -> CircuitBreakerState -> CircuitBreakerState
updateState now state =
  let -- Clean old failures outside monitoring window
      windowSeconds = fromIntegral (cbMonitoringWindow (cbOptions state)) / 1000.0 :: NominalDiffTime
      recentFailures = filter (\f -> diffUTCTime now (failureTimestamp f) < windowSeconds) (cbFailures state)
      
      -- Update state based on current state
      newState = case cbState state of
        CircuitClosed ->
          -- Check if we've exceeded failure threshold
          if length recentFailures >= cbFailureThreshold (cbOptions state) then
            CircuitOpen
          else
            CircuitClosed
        
        CircuitOpen ->
          -- Check if reset timeout has passed
          if now >= cbNextAttempt state then
            CircuitHalfOpen
          else
            CircuitOpen
        
        CircuitHalfOpen ->
          -- Stay in half-open until success or failure
          CircuitHalfOpen
      
      -- Update next attempt time if opening circuit
      newNextAttempt = if newState == CircuitOpen && cbState state == CircuitClosed then
        let resetTimeoutSeconds = fromIntegral (cbResetTimeout (cbOptions state)) / 1000.0 :: NominalDiffTime
        in addUTCTime resetTimeoutSeconds now
      else
        cbNextAttempt state
  in state
    { cbState = newState
    , cbFailures = recentFailures
    , cbNextAttempt = newNextAttempt
    }

-- | Record a failure
recordFailure :: UTCTime -> CircuitBreakerState -> CircuitBreakerState
recordFailure now state =
  state { cbFailures = FailureRecord now : cbFailures state }

-- | Record a success (reset failures if half-open)
recordSuccess :: UTCTime -> CircuitBreakerState -> CircuitBreakerState
recordSuccess now state =
  if cbState state == CircuitHalfOpen then
    state { cbState = CircuitClosed, cbFailures = [] }
  else
    -- Clean old failures but keep recent ones
    let windowSeconds = fromIntegral (cbMonitoringWindow (cbOptions state)) / 1000.0 :: NominalDiffTime
        recentFailures = filter (\f -> diffUTCTime now (failureTimestamp f) < windowSeconds) (cbFailures state)
    in state { cbFailures = recentFailures }

-- ════════════════════════════════════════════════════════════════════════════
--                                          // circuit // breaker // operations
-- ════════════════════════════════════════════════════════════════════════════

-- | Create a circuit breaker instance
-- 
-- System F/Omega proof: Explicit type IO (CircuitBreaker a)
createCircuitBreaker :: forall a. IO a -> CircuitBreakerOptions -> IO (CircuitBreaker a)
createCircuitBreaker fn options = do
  now <- getCurrentTime
  let safeOptions = CircuitBreakerOptions
        { cbFailureThreshold = if cbFailureThreshold options > 0 then cbFailureThreshold options else 5
        , cbResetTimeout = if cbResetTimeout options > 0 && isFinite (fromIntegral (cbResetTimeout options) :: Double) 
                          then cbResetTimeout options else 60000
        , cbMonitoringWindow = if cbMonitoringWindow options > 0 && isFinite (fromIntegral (cbMonitoringWindow options) :: Double)
                              then cbMonitoringWindow options else 60000
        , cbOnOpen = cbOnOpen options
        , cbOnClose = cbOnClose options
        , cbOnHalfOpen = cbOnHalfOpen options
        }
      initialState = CircuitBreakerState
        { cbState = CircuitClosed
        , cbFailures = []
        , cbNextAttempt = now
        , cbOptions = safeOptions
        }
  stateVar <- newMVar initialState
  return (CircuitBreaker fn stateVar)

-- | Execute the function with circuit breaker protection
-- 
-- System F/Omega proof: Explicit type IO (Either Text a)
execute :: forall a. CircuitBreaker a -> IO (Either Text a)
execute breaker = do
  now <- getCurrentTime
  state <- readMVar (cbStateVar breaker)
  
  let updatedState = updateState now state
  
  if cbState updatedState == CircuitOpen then
    -- Circuit is open, reject request
    let resetTimeoutSeconds = fromIntegral (cbResetTimeout (cbOptions updatedState)) / 1000.0 :: NominalDiffTime
        nextAttemptTime = addUTCTime resetTimeoutSeconds now
        errorMsg = "Circuit breaker is OPEN. Service unavailable. Retry after " <> T.pack (show nextAttemptTime)
    in return (Left errorMsg)
  else do
    -- Execute function
    result <- catch @SomeException (Right <$> cbFunction breaker) (\(e :: SomeException) -> return (Left (T.pack (show e))))
    
    modifyMVar_ (cbStateVar breaker) $ \s -> do
      now' <- getCurrentTime
      let s' = updateState now' s
      case result of
        Right _ -> do
          let s'' = recordSuccess now' s'
          -- Call onClose callback if transitioning from half-open to closed
          if cbState s' == CircuitHalfOpen && cbState s'' == CircuitClosed then
            case cbOnClose (cbOptions s'') of
              Just callback -> callback
              Nothing -> return ()
          else
            return ()
          return s''
        Left _ -> do
          let s'' = recordFailure now' s'
          -- Call onOpen callback if transitioning from closed to open
          if cbState s' == CircuitClosed && cbState s'' == CircuitOpen then
            case cbOnOpen (cbOptions s'') of
              Just callback -> callback
              Nothing -> return ()
          -- Call onHalfOpen callback if transitioning from open to half-open
          else if cbState s' == CircuitOpen && cbState s'' == CircuitHalfOpen then
            case cbOnHalfOpen (cbOptions s'') of
              Just callback -> callback
              Nothing -> return ()
          else
            return ()
          return s''
    
    return result

-- | Reset circuit breaker to closed state
reset :: CircuitBreaker a -> IO ()
reset breaker = do
  now <- getCurrentTime
  modifyMVar_ (cbStateVar breaker) $ \s -> do
    let s' = s { cbState = CircuitClosed, cbFailures = [], cbNextAttempt = now }
    case cbOnClose (cbOptions s') of
      Just callback -> callback
      Nothing -> return ()
    return s'

-- | Get current circuit state
getCurrentState :: CircuitBreaker a -> IO CircuitState
getCurrentState breaker = do
  now <- getCurrentTime
  state <- readMVar (cbStateVar breaker)
  let updatedState = updateState now state
  modifyMVar_ (cbStateVar breaker) $ \_ -> return updatedState
  return (cbState updatedState)

-- | Check if circuit is open
isOpen :: CircuitBreaker a -> IO Bool
isOpen breaker = (== CircuitOpen) <$> getCurrentState breaker

-- | Check if circuit is closed
isClosed :: CircuitBreaker a -> IO Bool
isClosed breaker = (== CircuitClosed) <$> getCurrentState breaker

-- | Check if circuit is half-open
isHalfOpen :: CircuitBreaker a -> IO Bool
isHalfOpen breaker = (== CircuitHalfOpen) <$> getCurrentState breaker

-- | Get failure count in current monitoring window
getFailureCount :: CircuitBreaker a -> IO Int
getFailureCount breaker = do
  now <- getCurrentTime
  state <- readMVar (cbStateVar breaker)
  let updatedState = updateState now state
  modifyMVar_ (cbStateVar breaker) $ \_ -> return updatedState
  return (length (cbFailures updatedState))

File: ui/src/utils/circuitBreaker.ts
/**
 * Circuit Breaker Utility
 *
 * Implements circuit breaker pattern to prevent cascading failures.
 * When a service is failing, the circuit breaker "opens" and stops
 * making requests until the service recovers.
 *
 * SECURITY: Prevents cascading failures and resource exhaustion.
 */

export type CircuitState = "closed" | "open" | "half-open";

export interface CircuitBreakerOptions {
  /** Number of failures before opening circuit (default: 5) */
  failureThreshold?: number;
  /** Time in milliseconds before attempting to close circuit (default: 60000) */
  resetTimeout?: number;
  /** Time window in milliseconds for counting failures (default: 60000) */
  monitoringWindow?: number;
  /** Optional callback when circuit opens */
  onOpen?: () => void;
  /** Optional callback when circuit closes */
  onClose?: () => void;
  /** Optional callback when circuit half-opens */
  onHalfOpen?: () => void;
}

const DEFAULT_OPTIONS: Required<
  Omit<CircuitBreakerOptions, "onOpen" | "onClose" | "onHalfOpen">
> = {
  failureThreshold: 5,
  resetTimeout: 60000, // 1 minute
  monitoringWindow: 60000, // 1 minute
};

interface FailureRecord {
  timestamp: number;
}

/**
 * Circuit Breaker class
 *
 * @example
 * ```typescript
 * const breaker = new CircuitBreaker(
 *   () => fetch('/api/data'),
 *   { failureThreshold: 3, resetTimeout: 30000 }
 * );
 *
 * try {
 *   const result = await breaker.execute();
 * } catch (error) {
 *   if (breaker.state === 'open') {
 *     // Circuit is open, service is down
 *   }
 * }
 * ```
 */
export class CircuitBreaker<T> {
  private state: CircuitState = "closed";
  private failures: FailureRecord[] = [];
  private nextAttempt: number = 0;
  private readonly options: Required<
    Omit<CircuitBreakerOptions, "onOpen" | "onClose" | "onHalfOpen">
  >;
  private readonly onOpen?: () => void;
  private readonly onClose?: () => void;
  private readonly onHalfOpen?: () => void;

  constructor(
    private readonly fn: () => Promise<T>,
    options: CircuitBreakerOptions = {},
  ) {
    this.options = {
      failureThreshold: options.failureThreshold ?? DEFAULT_OPTIONS.failureThreshold,
      resetTimeout: options.resetTimeout ?? DEFAULT_OPTIONS.resetTimeout,
      monitoringWindow:
        options.monitoringWindow ?? DEFAULT_OPTIONS.monitoringWindow,
    };
    this.onOpen = options.onOpen;
    this.onClose = options.onClose;
    this.onHalfOpen = options.onHalfOpen;
  }

  /**
   * Get current circuit state
   */
  get currentState(): CircuitState {
    this.updateState();
    return this.state;
  }

  /**
   * Check if circuit is open (blocking requests)
   */
  get isOpen(): boolean {
    return this.currentState === "open";
  }

  /**
   * Check if circuit is closed (allowing requests)
   */
  get isClosed(): boolean {
    return this.currentState === "closed";
  }

  /**
   * Check if circuit is half-open (testing recovery)
   */
  get isHalfOpen(): boolean {
    return this.currentState === "half-open";
  }

  /**
   * Update circuit state based on current conditions
   */
  private updateState(): void {
    const now = Date.now();

    // Clean old failures outside monitoring window
    this.failures = this.failures.filter(
      (f) => now - f.timestamp < this.options.monitoringWindow,
    );

    switch (this.state) {
      case "closed":
        // Check if we've exceeded failure threshold
        if (this.failures.length >= this.options.failureThreshold) {
          this.state = "open";
          this.nextAttempt = now + this.options.resetTimeout;
          this.onOpen?.();
        }
        break;

      case "open":
        // Check if reset timeout has passed
        if (now >= this.nextAttempt) {
          this.state = "half-open";
          this.onHalfOpen?.();
        }
        break;

      case "half-open":
        // Stay in half-open until success or failure
        break;
    }
  }

  /**
   * Record a failure
   */
  private recordFailure(): void {
    this.failures.push({ timestamp: Date.now() });
  }

  /**
   * Record a success (reset failures)
   */
  private recordSuccess(): void {
    if (this.state === "half-open") {
      this.state = "closed";
      this.failures = [];
      this.onClose?.();
    } else {
      // Clean old failures but keep recent ones
      const now = Date.now();
      this.failures = this.failures.filter(
        (f) => now - f.timestamp < this.options.monitoringWindow,
      );
    }
  }

  /**
   * Execute the function with circuit breaker protection
   *
   * @throws Error if circuit is open or function fails
   */
  async execute(): Promise<T> {
    this.updateState();

    if (this.state === "open") {
      throw new Error(
        `Circuit breaker is OPEN. Service unavailable. Retry after ${new Date(this.nextAttempt).toISOString()}`,
      );
    }

    try {
      const result = await this.fn();
      this.recordSuccess();
      return result;
    } catch (error) {
      this.recordFailure();
      this.updateState();
      throw error;
    }
  }

  /**
   * Reset circuit breaker to closed state
   */
  reset(): void {
    this.state = "closed";
    this.failures = [];
    this.nextAttempt = 0;
    this.onClose?.();
  }

  /**
   * Get failure count in current monitoring window
   */
  getFailureCount(): number {
    this.updateState();
    return this.failures.length;
  }
}

/**
 * Create a circuit breaker instance
 *
 * @example
 * ```typescript
 * const breaker = createCircuitBreaker(
 *   () => apiCall(),
 *   { failureThreshold: 3 }
 * );
 * ```
 */
export function createCircuitBreaker<T>(
  fn: () => Promise<T>,
  options?: CircuitBreakerOptions,
): CircuitBreaker<T> {
  return new CircuitBreaker(fn, options);
}
